### ISSUE TITLE
Fitting algorithms 

### ISSUE THREAD

**Issue by tjof2** [hyperspy/hyperspy#1098]
Looking through `model.py` I don't understand the documentation and usage of the various fitting algorithms available, particularly from `scipy.optimize`. This was also mentioned in #946.

I take issue with the use of the word `robust` here:

> "fmin" performs curve fitting using a downhill simplex algorithm.
> It is **less robust** than the Levenberg-Marquardt based optimizers,
> but, at present, it is the only one that support maximum likelihood
> optimization for Poissonian noise.

**Robust** in what sense? MLE will achieve the Cramer-Rao lower bound, and it will be equivalent to least-squares in the presence of Gaussian noise. 

If it means **robust** in the sense of it doesn't find a good minimum, why are we importing things like Conjugate-Gradient, BFGS and L-BFGS at the top of the file but not using/referring to them? Nelder-Mead can actually perform pretty well in some cases, especially if derivatives aren't available or are badly-behaved when calculated numerically using e.g. finite differences. In any case, CG and BFGS should perform as well as LM but with the bonus of working with functions that don't use an _L2_-norm.

Building on the point about all the other algorithms in `scipy.optimize`, this isn't consistent (line 710):
```
if fitter not in ("mpfit", "tnc", "l_bfgs_b"):
     raise NotImplementedError("Bounded optimization is only available for the mpfit optimizer.")
```

I don't completely follow the use of bound constraints in `mpfit`. I'm sure they're correct as it's been around for 30 years or more, but I would have thought `scipy.optimize` has some more modern methods for bound constraints. Also, does this mean one cannot combine Poisson MLE with bound constraints? Surely Poisson noise is a case where the model is guaranteed to be positive!

In summary, I think it would be good to:

- Make it easier to use the different algorithms in `scipy.optimize`, or at least tidy up the inconsistencies.
- Clarify the documentation, particularly around the use of "robust".

**Comment by tjof2** [comment]
Also, in #946 @francisco-dlp mentions:

> I vaguely remember testing mpfit vs leastsq when I was considering it for inclusion in HyperSpy and finding that they were converging to the same values when bounding was not necessary. 

If I understand correctly, `mpfit` builds upon `lmdif` from MINPACK, while [`scipy.optimize.leastsq`](http://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.leastsq.html) is a wrapper around...`lmdif` from MINPACK. 

Does `mpfit` offer any particular benefit over `scipy.leastsq`, and if so, is it superseded by [`scipy.optimize.least_squares`](http://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.least_squares.html) which implements trust regions as well as robust loss functions (e.g. L1-norm).

**Comment by francisco-dlp** [comment]
Good point @tjof2. I wasn't aware of the new bounded least squares functionality in Scipy. We should definitely add support for it and possibly deprecate mpfit. 

> Surely Poisson noise is a case where the model is guaranteed to be positive!

It's the data that needs to be positive not the parameter values.

A good summary on when each of the algorithms in scipy.optimize is appropriate can be found in the [Scipy Lectures Notes](http://www.scipy-lectures.org/advanced/mathematical_optimization/index.html). The code and documentation in `Model.fit` is very old, probably written in late 2007 and haven't been updated much since then. If now there are other bounded optimizers in Scipy and you feel like adding support for them you are more than welcome. Regarding ML vs WLS, in my experience ML is only clearly better for Poisson noise when the SNR is very low. This is often the case in EDX, so it'll be indeed interesting to have a bounded ML optimiser.

**Comment by tjof2** [comment]
> It's the data that needs to be positive not the parameter values.

Sorry yes my mistake - though the model function value should also be positive - but the bounds are on parameters not the function value.

What's the deal on being able to pass analytic gradients and Hessians, if available?

**Comment by francisco-dlp** [comment]
Regarding ML vs WLS, yes, that's what I meant. Of course ML provides the best result for Poisson data everything else being equal. Currently the ML implementation in hyperspy is fairly incomplete e.g.  there is no estimation of the accuracy of the parameter value. This is because I had little motivation to work on these features because my EELS data wouldn't benefit much from it and nobody else has contributed to this part of the code. Again EDX people (among others) should be interested on this.

By passing `grad=True` it is possible to use analytical gradients for the algorithms that support it. The code that computes the jacobian is also very old and there is probably a lot of room for improvement and simplification. However, it should work. It'll be nice to compute the hessian too analytically. Since we have the `Expression` components, this doesn't involve manually defining the second derivatives any more.

**Comment by tjof2** [comment]
There should be a way to work in estimation of the Fisher information matrix when gradients are available and hence get the parameter covariances out by inverting.

This might be a few separate strands then: 

- Investigating `scipy.optimize.least_squares` further and seeing if it means `mpfit` can be deprecated.
- Making it clearer on how to access the other `scipy.optimize` algorithms, in particular CG and BFGS/L-BFGS as alternatives for Nelder-Mead
- Improving the MLE for Poisson data to get estimates of parameter variance.

**Comment by francisco-dlp** [comment]
Sounds like a great plan.
